lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+            }
+            uStep += pixelWidth;
+        }
+    }
+}
+
+function computeAmbientOcclusionAt(position, normal, numberSamples, sqrtNumberSamples,
+    triangles, nearCull, rayDepth, aoBuffer, aoBufferIndex) {
+    for (var j = 0; j < numberSamples; j++) {
+        var sampleRay = generateJitteredRay(position, normal, j, sqrtNumberSamples);
+        aoBuffer.count[aoBufferIndex]++;
+        var nearestIntersect = naiveRaytrace(triangles, sampleRay, nearCull);
+
+        if (nearestIntersect < rayDepth) {
+            aoBuffer.samples[aoBufferIndex] += 1.0;
+        }
+    }
+}
+
+function naiveRaytrace(triangleSoup, ray, nearCull) {
+    // Check ray against every triangle in the soup. return the nearest intersection.
+    var minIntersect = Number.POSITIVE_INFINITY;
+    for (var triangleSoupIndex = 0; triangleSoupIndex < triangleSoup.length; triangleSoupIndex++) {
+        var positions = triangleSoup[triangleSoupIndex].positions;
+        var distance = rayTriangle(ray, positions[0], positions[1], positions[2], false);
+        if (defined(distance) && distance > nearCull) {
+            minIntersect = Math.min(distance, minIntersect);
+        }
+    }
+    return minIntersect;
+}
+
+function generateJitteredRay(position, normal, sampleNumber, sqrtNumberSamples) {
+    // Stratified (jittered) Sampling with javascript's own rand function
+    // Based on notes here: http://graphics.ucsd.edu/courses/cse168_s14/ucsd/CSE168_11_Random.pdf
+
+    // Produces samples based on a grid of dimension sqrtNumberSamples x sqrtNumberSamples
+    var cellWidth = 1.0 / sqrtNumberSamples;
+    var s = (sampleNumber % sqrtNumberSamples) * cellWidth + (Math.random() / sqrtNumberSamples);
+    var t = Math.floor(sampleNumber / sqrtNumberSamples) * cellWidth + (Math.random() / sqrtNumberSamples);
+
+    // Generate ray on a y-up hemisphere with cosine weighting (more rays around the normal)
+    var u = 2.0 * Math.PI * s;
+    var v = Math.sqrt(1.0 - t);
+
+    var randomDirection = scratchRay.direction;
+    randomDirection.x = v * Math.cos(u);
+    randomDirection.y = t;
+    randomDirection.z = v * Math.sin(u);
+
+    // Orient with texel's normal in world space
+    var theta = Math.acos(normal.y); // dot product of normal with y-up is normal.y
+    var axis = Cartesian3.cross(randomDirection, normal, cartesian3Scratch1);
+    var rotation = Quaternion.fromAxisAngle(axis, theta, quaternionScratch);
+    var matrix = Matrix3.fromQuaternion(rotation, matrix3Scratch);
+
+    scratchRay.origin = position;
+    scratchRay.direction = Matrix3.multiplyByVector(matrix, randomDirection, scratchRay.direction);
+    return scratchRay;
+}
+
+// Borrowed straight from Cesium/Source/Core/IntersectionTests
+var scratchEdge0 = new Cartesian3();
+var scratchEdge1 = new Cartesian3();
+var scratchPVec = new Cartesian3();
+var scratchTVec = new Cartesian3();
+var scratchQVec = new Cartesian3();
+
+function rayTriangle(ray, p0, p1, p2, cullBackFaces) {
```

Comment:
The parametric version currently isn't exported, is there another way to access it outside `IntersectionTests.js`? I wanted to avoid having to convert from `Cartesian3` back to parametric distance since the ray-triangle test is so hot.